Add Thai word list from ICU BreakIterator dictionary

pavaris-pm opened the pull request:
### What does this changes

@wannaphong @bact  from issue #877 since ICU are included to almost all web browser,  i've added ICU dictionary to PyThaiNLP where file of ICU dictionary are named as `icubrk_th.txt` and their python file to load the corpus are named as `thai_icu.py` krub.

Will resolve #877 

### Your checklist for this pull request
🚨Please review the [guidelines for contributing](https://github.com/PyThaiNLP/pythainlp/blob/dev/CONTRIBUTING.md) to this repository.

- [x] Passed code styles and structures
- [ ] Passed code linting checks and unit test

pavaris-pm committed:
add thai ICU corpus

pep8speaks commented:
Hello @pavaris-pm! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:









There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2023-12-06 10:17:06 UTC

pavaris-pm committed:
fix pep8

wannaphong commented:
Hello! Thank you for your pull request. Can you add filter the word that start with `#`?

bact committed:
Add SPDX tags to thai_icu.txt

SPDX-License-Identifier: Unicode-DFS-2016

pavaris-pm commented:
> Hello! Thank you for your pull request. Can you add filter the word that start with `#`?

Sure. Did you mean add a parameters for user to control whether to return a corpus with the text starts with # or not right? by **True** if you want a returned corpus including words starts with #, and returned the corpus with filtered out word starts with # (no word start with # in corpus) otherwise.

wannaphong commented:
> > Hello! Thank you for your pull request. Can you add filter the word that start with `#`?
> 
> Sure. Did you mean add a parameters for user to control whether to return a corpus with the text starts with # or not right? by **True** if you want a returned corpus including words starts with #, and returned the corpus with filtered out word starts with # (no word start with # in corpus) otherwise.

Yes 👍

bact committed:
Sort imports in __init__.py

bact commented on pythainlp/corpus/thai_icu.py:
```diff
@@ -0,0 +1,30 @@
+    """
+    Return a frozenset of words from the International Components for Unicode (ICU) dictionary.
+
+    The data is at pythainlp/corpus/thai_icu.txt
+    The word list has beed prepared by the code at:
+    https://github.com/unicode-org/icu/blob/main/icu4c/source/data/brkitr/dictionaries/thaidict.txt
```
Do you mind to move these info to the corpus_license.md instead please?

https://github.com/PyThaiNLP/pythainlp/blob/dev/pythainlp/corpus/corpus_license.md

pavaris-pm replied:
Roger that 🫡

pavaris-pm replied:
done !

bact reviewed (changes requested):
Once license info has moved to corpus_license.md AND the comment lines are properly discarded, I can merge this.

bact commented:
> > Hello! Thank you for your pull request. Can you add filter the word that start with `#`?
> 
> Sure. Did you mean add a parameters for user to control whether to return a corpus with the text starts with # or not right? by **True** if you want a returned corpus including words starts with #, and returned the corpus with filtered out word starts with # (no word start with # in corpus) otherwise. 

I think we can do this in `get_corpus()`.

Maybe add the boolean parameter `discard_comments` to `get_corpus()`? 
The default is probably `False`.


Or, we can utilize the existing Python standard library `shlex` for this. shlex will ignore comment lines when it gets its input.

https://docs.python.org/3/library/shlex.html

pavaris-pm committed:
add comment filtering and update corpus license

pavaris-pm commented:
> > > Hello! Thank you for your pull request. Can you add filter the word that start with `#`?
> > 
> > 
> > Sure. Did you mean add a parameters for user to control whether to return a corpus with the text starts with # or not right? by **True** if you want a returned corpus including words starts with #, and returned the corpus with filtered out word starts with # (no word start with # in corpus) otherwise.
> 
> I think we can do this in `get_corpus()`.
> 
> Maybe add the boolean parameter `discard_comments` to `get_corpus()`? The default is probably `False`.
> 
> Or, we can utilize the existing Python standard library `shlex` for this. shlex will ignore comment lines when it gets its input.
> 
> https://docs.python.org/3/library/shlex.html

@bact @wannaphong i already add comment filtering by adding a new parameters named `discard_comments` where the default value is set to be `False`. You can review the code from the latest commit krub

pavaris-pm commented on pythainlp/corpus/corpus_license.md:
```diff
@@ -53,6 +53,13 @@ https://creativecommons.org/licenses/by/4.0/
+## Thai International Components for Unicode (ICU)
+The data is at pythainlp/corpus/thai_icu.txt
+The word list has beed prepared by the code at:
+https://github.com/unicode-org/icu/blob/main/icu4c/source/data/brkitr/dictionaries/thaidict.txt
```
@bact already moved the text in `thai_icu` to `corpus_license.md` already krub.

pavaris-pm reviewed (commented):
thanks for help me sorting it alphabetically krub 👍🏻

pavaris-pm committed:
fix pep8

pavaris-pm committed:
fix pep8 (trailing whitespaces)

pavaris-pm committed:
fix bug in thai_icu

pavaris-pm committed:
fix typo

pavaris-pm commented:
> > > Hello! Thank you for your pull request. Can you add filter the word that start with `#`?
> > 
> > 
> > Sure. Did you mean add a parameters for user to control whether to return a corpus with the text starts with # or not right? by **True** if you want a returned corpus including words starts with #, and returned the corpus with filtered out word starts with # (no word start with # in corpus) otherwise.
> 
> I think we can do this in `get_corpus()`.
> 
> Maybe add the boolean parameter `discard_comments` to `get_corpus()`? The default is probably `False`.
> 
> Or, we can utilize the existing Python standard library `shlex` for this. shlex will ignore comment lines when it gets its input.
> 
> https://docs.python.org/3/library/shlex.html

@bact  @wannaphong I've made some experiment to test the `discard_comments` parameters and fix some bugs from it. Now it works perfectly. feel free to review from now on krub. It's done 💯

wannaphong committed:
Add more get_corpus docs

wannaphong reviewed (approved):
It look great for me.

bact committed:
Add license for Thai dict from ICU

Filename: icubrk_th.txt
License: Unicode-DFS-2016

bact committed:
Rename thai_icu.txt to icubrk_th.txt

bact committed:
Adjust comment discard method in get_corpus()

bact committed:
Update and rename thai_icu.py to icu.py

Also rename `thai_icu()` to `thai_icu_words()` to make it more explicit and consistent with others, like: `thai_orst_words()`

bact committed:
Update __init__.py

Change thai_icu to thai_icu_words

bact committed:
Update core.py

bact committed:
Update test_corpus.py

bact commented on pythainlp/corpus/core.py:
```diff
@@ -70,7 +70,10 @@ def path_pythainlp_corpus(filename: str) -> str:
-def get_corpus(filename: str, as_is: bool = False) -> Union[frozenset, list]:
+def get_corpus(filename: str,
+               as_is: bool = False,
+               comments: bool = True
```
I have changed this to `comments` instead of `discard_comments` (as I suggested earlier) to avoid double negation. 

The semantic now is:
- if `comments` = True, then keep comments
- if `comments` = False, then discard comments

bact commented on pythainlp/corpus/core.py:
```diff
@@ -93,26 +100,61 @@ def get_corpus(filename: str, as_is: bool = False) -> Union[frozenset, list]:
 
+    if not comments:
+        # take only text before character '#'
+        lines = [line.split("#", 1)[0] for line in lines]
```
This will allowed the comment to be at any position of the line.

bact reviewed (approved):
Approved.

Few modification to `get_corpus()` to make the code more generic.

I have changed the module/function name to
- `corpus.icu` instead of `corpus.thai_icu` - to make the module name more generic
- `thai_icu_words` instead of `thai_icu` - to make the function name inline with `thai_words` and `thai_orst_words`

So when import, it will be like:
```python
from pythainlp.corpus.icu import thai_icu_words
```

Note: I will also after this rename the wikipedia (#869) and volubilis (#870) corpora as well, to make them more consistent:

So instead of having:
```python
from pythainlp.corpus.volubilis import volubilis
from pythainlp.corpus.wikipedia_titles import wikipedia_titles
```

we should have:
```python
from pythainlp.corpus.volubilis import thai_volubilis_words
from pythainlp.corpus.wikipedia import thai_wikipedia_titles
```

bact commented:
Merged thank you.
